### ISSUE TITLE
SMTPSenderFactory always fails on retry

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3638]
|<img alt="atsuoi's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| atsuoi reported|
|-|-|
|Trac ID|trac#3638|
|Type|defect|
|Created|2009-01-31 17:29:01Z|
|Branch|https://github.com/twisted/twisted/tree/smtpsender-retry-3638-2|

If SMTP session was failed after file transfer started, then retrying will always fail with following error:

  Failure instance: Traceback (failure with no frames): <class 'twisted.mail.smtp.SMTPDeliveryError'>: No recipients accepted

I attached a patch to fix problem. With this patch, the file object passed to SMTPSenderFactory should have seek() method.

Attachments:

* [smtp.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/02b/02b1a4b2bd92d82a564ad9fca51573a3d56fd09d/76cae3b58b8afd675c1a965daf7ccce1090ffa7f.diff) (706 bytes) - added by atsuoi on 2009-01-31 17:29:23Z - 
* [test_smtp.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/02b/02b1a4b2bd92d82a564ad9fca51573a3d56fd09d/978d9a5d2e9169d3e3e2b0c58fa0b849df2a4fc6.diff) (1855 bytes) - added by atsuoi on 2009-02-01 05:23:52Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__3638 3638
type__defect defect
reporter__atsuoi atsuoi
priority__normal normal
milestone__ 
branch__branches_smtpsender_retry_3638_2 branches/smtpsender-retry-3638-2
branch_author__atsuoi__exarkun atsuoi, exarkun
status__closed closed
resolution__fixed fixed
component__mail mail
keywords__ 
time__1233422941000000 1233422941000000
changetime__1233865142000000 1233865142000000
version__None None
owner__ 
cc__exarkun cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun set owner to ascii|
|-|-|

Thanks for the patch.  Can you also attach a unit test for the behavior being fixed?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_2"></a>|@glyph set owner to atsuoi|
|-|-|

Fixing (apparent?) mis-assignment.

**Comment by twisted-trac** [comment]
|<img alt="atsuoi's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|atsuoi commented|
|-|-|

Add an unit test. I'm not familiar with Twisted's unit test, so the test may contain uncessecary codes. I'd appreciate to know how I can improve this test.

**Comment by twisted-trac** [comment]
|<img alt="atsuoi's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|atsuoi removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

  1. Generally, tests shouldn't do things which require superuser privileges.  Binding to port 25 will fail if the test suite is run as a normal user, which is almost always how it's run.  Even if the suite is run as a superuser, it will fail if there's a real mail server listening.  `reactor.listenTCP(0, ...)` will request that an unused port is used.  The returned port object will tell you what port number was actually bound - `port.getHost().port`.  This is generally the preferred thing to do if it's necessary to bind a port in a unit test.
  1. All classes and methods should have a docstring.
  1. In the Twisted naming convention, variables and attributes use the `camelCase` convention.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

(In [26174]) Branching to 'smtpsender-retry-3638'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

In addition to fixing the above points, I've also changed the test to run against a real SMTP server.  I was tempted to leave the fake SMTP server in place, but the complexity of the interaction, particular the timing of certain events, tipped me over.  It's a bit more code, but not much, and it should be a better test.

[Build Results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/smtpsender-retry-3638)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_9"></a>|@thijstriemstra set owner to @exarkun|
|-|-|

Make sure to update the copyright headers for 2009 in the files you modified.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun removed owner|
|-|-|

This branch touches the same files as [#3641](https://github.com/twisted/twisted/issues/3641), which updates the copyright date.  No point changing them here too.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_11"></a>|@therve set owner to @exarkun|
|-|-|

Cool, +1.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

(In [26190]) Branching to 'smtpsender-retry-3638-2'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun set status to `closed`|
|-|-|

(In [26193]) Merge smtpsender-retry-3638-2


Author: atsuoi, exarkun
Reviewer: therve
Fixes: [#3638](https://github.com/twisted/twisted/issues/3638)

Fix the retry support in `SMTPSenderFactory` so that if one or more recipient
addresses is accepted or if part of the message body file is read and sent to
the remote MTA that data is still used on subsequent delivery attempts.  The
previous behavior could exhaust the recipient address iterator on the first
failed attempt or move the position pointer in the message body file causing
subsequent deliver attempts to operate using invalid data.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
